### PR TITLE
Add Animals

### PR DESCRIPTION
**Pull request by michelfinley**
## Summary

This PR implements animals as described in issue #64 .

This PR is still work in progress, changes made as of now:
The Entity class and its children have been refactored to allow for player-unlike AI-controlled Entities.

## Checklist

- [x] I have tested this change locally and it works as expected.
- [x] I have made sure that the code follows the formatting and style guidelines of the project.

## Labels
https://github.com/sloukit/pydew-valley-uzh/labels/game-ai https://github.com/sloukit/pydew-valley-uzh/labels/game-gameplay https://github.com/sloukit/pydew-valley-uzh/labels/priority%3A%20medium https://github.com/sloukit/pydew-valley-uzh/labels/type%3A%20feature

### CODE REVIEW

File: src/npc/__init__.py
Comment (fortwoone) [inline src/npc/__init__.py:1]: I’d probably put a docstring in this file, to describe the package’s content

File: src/npc/bases/__init__.py
Comment (fortwoone) [inline src/npc/bases/__init__.py:1]: Same comment as `src.npc.__init__`

File: src/npc/behaviour/__init__.py
Comment (fortwoone) [inline src/npc/behaviour/__init__.py:1]: Same comment as before

File: src/npc/behaviour/ai_behaviour_base.py [hunk 1]
```diff
@@ -7,37 +7,35 @@
 from pathfinding.core.grid import Grid
 from pathfinding.finder.a_star import AStarFinder
 
-from src.overlay.soil import SoilLayer
 from src.sprites.entity import Entity
 
 
 # TODO: Refactor NPCState into Entity.state (maybe override Entity.get_state())
-class NPCState(IntEnum):
+class AIState(IntEnum):
```
Comment (fortwoone) [inline src/npc/behaviour/ai_behaviour_base.py:14]: This enum could be used for the player as well, not just CPU-controlled entities

File: src/npc/behaviour/chicken_behaviour_tree.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+
+@dataclass
+class ChickenBehaviourTreeContext(Context):
+    chicken: ChickenBase
+
+
+class ChickenBehaviourTree:
+    """
+    Group of classes used for Chicken behaviour.
```
Comment (fortwoone) [inline src/npc/behaviour/chicken_behaviour_tree.py:20]: All instances of "Group of **classes**" should be replaced by "group of **methods**", since it’s functions you use, not classes

File: src/screens/level.py [hunk 3]
```diff
@@ -1,40 +1,42 @@
...
+from src.sprites.tree import Tree
+from src.support import map_coords_to_tile, load_data, resource_path
 
 
 class Level:
-    def __init__(self, game, switch, tmx_maps: MapDict, frames, sounds):
+    def __init__(self, switch, tmx_maps: MapDict, frames, sounds):
         # main setup
         self.display_surface = pygame.display.get_surface()
-        self.game = game
```
Comment (fortwoone) [inline src/screens/level.py:37]: Why did you remove this?

**Review by fortwoone — COMMENTED** [review]
Before I can approve or request changes, I have a few questions to ask

File: main.py [hunk 4]
```diff
@@ -45,7 +46,7 @@ def __init__(self):
         self.load_assets()
 
         # screens
-        self.level = Level(self, self.switch_state, self.tmx_maps, self.frames, self.sounds)
+        self.level = Level(self.switch_state, self.tmx_maps, self.frames, self.sounds)
```
[the remark this answers is not in the document]
Comment (michelfinley) [reply main.py:49]: Actually, I wasn't thinking about the dialogue manager when removing game as a parameter. I thought that it is usually desirable to remove too unspecific dependencies, and I couldn't find Level.game / Player.game being used anywhere, testing the dialogue feature must have been accidentally removed by a previous PR. I totally get your point though, should I add it back in? Or should I instead pass only the dialogue manager?

File: src/enums.py [hunk 5]
```diff
@@ -198,4 +198,16 @@ def as_nonseed_ir(self):
...
+class Direction(StrEnum):
+    UP = "up"
+    RIGHT = "right"
+    DOWN = "down"
+    LEFT = "left"
+
+
+class EntityState(StrEnum):
+    IDLE = "idle"
+    WALK = "walk"
```
[the remark this answers is not in the document]
Comment (michelfinley) [reply src/enums.py:213]: Oh, thanks for the catch 😅
I'll add it right away

File: src/npc/__init__.py
Comment (michelfinley) [reply src/npc/__init__.py:1]: All right, thank you

File: src/npc/behaviour/ai_behaviour_base.py (hunk 1 above)
Comment (michelfinley) [reply src/npc/behaviour/ai_behaviour_base.py:14]: You're right, I've been thinking about that too. Yesterday, inspired by raphaels PR, I played around with the entity animation system and changed the raw string references in Entity.frames to enum members. I think that could be used here instead of the AIState as well, although the current way NPC pathfinding works is still dependent on it being separate from the Entity.state. I'm working on changing that, thanks for pointing it out!

File: src/npc/behaviour/chicken_behaviour_tree.py (hunk 2 above)
Comment (michelfinley) [reply src/npc/behaviour/chicken_behaviour_tree.py:20]: You're right, I'll change that right away 😅

File: main.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (fortwoone) [reply main.py:49]: Actually, it’s probably better this way as it allows us to let NPCs open the text boxes themselves

**Review by fortwoone — CHANGES_REQUESTED** [review]
Actually, I just checked the presence of conflicts, so a rebase is required anyway

File: src/enums.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (michelfinley) [reply src/enums.py:213]: Huh, somehow my IDE and notepad++ both show me a newline at the end of that file, could it be that GitHub just doesn't display it?

[the remark this answers is not in the document]
Comment (fortwoone) [reply src/enums.py:213]: Hmm, may be GitHub then

File: src/npc/__init__.py
Comment (michelfinley) [reply src/npc/__init__.py:1]: Are the docstrings I just pushed fine as they are? Or should they contain more specific information about the documented module? This is the first time I'm writing module-level docstrings, so please be encouraged to be a little more critical.

File: main.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (michelfinley) [reply main.py:49]: Somehow the github website is lagging behind on my end, so I accidentally added the attribute again... Thank you for your review!

File: src/npc/__init__.py
Comment (fortwoone) [reply src/npc/__init__.py:1]: Yes, they’re fine! You only need to rebase your work onto the main branch now

Comment (michelfinley) [reply src/npc/__init__.py:1]: Done, thank you very much 😄

**Review by fortwoone — APPROVED** [review]
LGTM
